doc/man/66-instate.1.scd
66-instate(1)

# NAME

66-instate - displays the contents of the service's state file.

# SYNOPSYS

66-instate [ *-h* ] [ *-z* ] [ *-t* _tree_ ] [ *-l* ] _service_

# DESCRIPTION

*66-instate* displays the contents of the _service_'s state file.
This file are used internally by the 66 tools to know runtime _service_ information.
This tool is purely a debug tool used by developers.

# OPTIONS

*-h*
	Prints this help.

*-z*
	use color.

*-t* _tree_
	only searches the _service_ at the specified _tree_, when the same 
	_service_ may be enabled in more trees.

*-l*
	displays the contents of the associated logger's state file if any.


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 
    complete name of the frontend service file -- Refer to the frontend 
    documentation page.
    
    - New skeleton file: rc.shutdown.final. This skeleton file will be run
    at the very end of the shutdown procedure, after all processes have 
    been killed and all filesystems have been unmounted, just before the 
    system is rebooted or the power turned off.
    This script normally remains empty.
   
    - Extra tools has removed and provided as an independent programs at
    https://framagit.org/Obarun/66-tools.git except for 66-echo and 
    66-unmountall program which are a dependent part of 66.
   
    - The account to run the s6-log program at the associate service logger
    can be set at compilation time by the --with-s6-log-user=USER options.
    The default is root. Obviously, the @runas field at the [logger] section
    overwritte it.
    This options set the account to run the uncaught-logs too. Also,
    this option can be overwritten with the '-l' option at 66-boot invocation
    and the '-L' option at 66-scandir invocation. 
    
In 0.2.0.4
----------
	
	- Bugs fix on memory allocation
	- Respect policies decision of user: 
		Remove -m option from 66-boot on init file
	- Add -d feature to 66-scandir 

In 0.2.0.3
----------

	- 66-parser: Fix write of configuration file, add -c|C features 
	
In 0.2.0.2
----------

	- Fix the written of user configuration file 
	- Fix oneshot and bundle status check
	- Add -l features to 66-tree, rewrite tree_unsupervise function
	
In 0.2.0.0
----------

	- New tools:
		- 66-boot.
		- 66-scanctl.
		- 66-shutdown.
		- 66-shutdownd.
		- 66-hpr.
		- 66-env.
		- 66-parser.
		- 66-which.
		- 66-echo.
		- 66-unmountall.
		- execl-subuidgid.
	- New @key field and change on frontend file:
		- [Logger] section accept a @depends field, @timestamp accept none
		as value, readiness notification is a default.
		- @hiercopy in [main] allow to copy any file or directory coming
		from the directory of the service.
	- 66-envfile is now deprecated, use execl-envfile in replacement.
	- 66-enable: add -F, -c, -C features.
	- 66-svctl is now asynchrone to launch services.
	- Man pages are available.
	- Environment variables limitation: maximum 100 files by directory,
	each file cannot contain more than 4096 bytes or 50 variables.

src/66/66-instate.c
/* 
 * 66-state.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 */

#include <string.h>
#include <wchar.h>

#include <oblibs/log.h>
#include <oblibs/sastr.h>
#include <oblibs/string.h>
#include <oblibs/obgetopt.h>
#include <oblibs/types.h>

#include <skalibs/buffer.h>
#include <skalibs/types.h>
#include <skalibs/stralloc.h>
#include <skalibs/lolstdio.h>

#include <66/resolve.h>
#include <66/info.h>
#include <66/utils.h>
#include <66/constants.h>
#include <66/state.h>

#define MAXOPTS 7

static wchar_t const field_suffix[] = L" :" ;
static char fields[INFO_NKEY][INFO_FIELD_MAXLEN] = {{ 0 }} ;

#define USAGE "66-instate [ -h ] [ -z ] [ -t tree ] [ -l ] service"

static inline void info_help (void)
{
  static char const *help =
"66-instate <options> service \n"
"\n"
"options :\n"
"	-h: print this help\n"
"	-z: use color\n"
"	-t: only search service at the specified tree\n"
"	-l: prints information of the associated logger if exist\n"
;

 if (buffer_putsflush(buffer_1, help) < 0)
    log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
}

static void info_display_string(char const *field,char const *str)
{
	info_display_field_name(field) ;
	
	if (!*str)
	{
		if (!bprintf(buffer_1,"%s%s",log_color->warning,"None"))
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;		
	}
	else
	{
		if (!buffer_puts(buffer_1,str))
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	}
	if (buffer_putsflush(buffer_1,"\n") == -1) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
}

static void info_display_int(char const *field,unsigned int id)
{
	char *str = 0 ;
	char ival[UINT_FMT] ;
	ival[uint_fmt(ival,id)] = 0 ;
	str = ival ;
	
	info_display_string(field,str) ;		
}

int main(int argc, char const *const *argv) 
{
	size_t newlen = 0, pos = 0 ;
	int found = 0 ;
	uint8_t logger = 0 ;
	uid_t owner ;
	char ownerstr[UID_FMT] ;
	ss_resolve_t res = RESOLVE_ZERO ;
	stralloc satree = STRALLOC_ZERO ;
	stralloc src = STRALLOC_ZERO ;
	stralloc tmp = STRALLOC_ZERO ;
	ss_state_t sta = STATE_ZERO ;
	char const *tname = 0 ;
	char const *svname = 0 ;
	char const *ste = 0 ;
	
	log_color = &log_color_disable ;
	
	char buf[MAXOPTS][INFO_FIELD_MAXLEN] = {
		"Reload",
		"Init",
		"Unsupervise",
		"State",
		"Pid",
		"Name" ,
		"Real logger name" } ;
	
	PROG = "66-instate" ;
	{
		subgetopt_t l = SUBGETOPT_ZERO ;

		for (;;)
		{
			int opt = getopt_args(argc,argv, ">hzlt:", &l) ;
			if (opt == -1) break ;
			if (opt == -2) log_die(LOG_EXIT_USER,"options must be set first") ;
			switch (opt)
			{
				case 'h' : info_help(); return 0 ;
				case 'z' : log_color = !isatty(1) ? &log_color_disable : &log_color_enable ; break ;
				case 't' : tname = l.arg ; break ;
				case 'l' : logger = 1 ; break ;
				default : log_usage(USAGE) ; 
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	
	if (!argc) log_usage(USAGE) ;
	svname = *argv ;
		
	owner = getuid() ;
	size_t ownerlen = uid_fmt(ownerstr,owner) ;
	ownerstr[ownerlen] = 0 ;
		
	if (!set_ownersysdir(&src,owner)) log_dieusys(LOG_EXIT_SYS, "set owner directory") ;
	if (!auto_stra(&src,SS_SYSTEM)) log_die_nomem("stralloc") ;
	
	if (!scan_mode(src.s,S_IFDIR))
	{
		log_info("no tree exist yet") ;
		goto freed ;
	}
	
	if (!auto_stra(&src,"/")) log_die_nomem("stralloc") ;
	newlen = src.len ;
	
	if (!tname)
	{	
		if (!stralloc_0(&src) ||
		!stralloc_copy(&tmp,&src)) log_die_nomem("stralloc") ;
		
		if (!sastr_dir_get(&satree, src.s,SS_BACKUP+1, S_IFDIR)) 
			log_dieu(LOG_EXIT_SYS,"get tree from directory: ",src.s) ;
		
		if (satree.len)
		{
			for(pos = 0 ; pos < satree.len ; pos += strlen(satree.s + pos) + 1)
			{
				tmp.len = newlen ;
				char *name = satree.s + pos ;
				
				if (!auto_stra(&tmp,name,SS_SVDIRS)) log_die_nomem("stralloc") ;
				if (ss_resolve_check(tmp.s,svname))
				{
					if (!found)
						if (!stralloc_copy(&src,&tmp)) log_die_nomem("stralloc") ;
					found++ ;
				}
			}
		}
		else 
		{
			log_info("no tree exist yet") ;
			goto freed ;
		}
	}
	else
	{
		if (!auto_stra(&src,tname,SS_SVDIRS)) log_die_nomem("stralloc") ;
		if (ss_resolve_check(src.s,svname)) found++;
	}

	if (!found)
	{
		log_die(LOG_EXIT_SYS,"unknown service: ",svname) ;
	
	}
	else if (found > 1)
	{
		log_die(LOG_EXIT_SYS,svname," is set on different tree -- please use -t options") ;
	}
	if (!ss_resolve_read(&res,src.s,svname)) log_dieusys(111,"read resolve file of: ",src.s,"/.resolve/",svname) ;
	
	info_field_align(buf,fields,field_suffix,MAXOPTS) ;
	
	ste = res.sa.s + res.state ;
	
	if (!ss_state_check(ste,svname)) log_diesys(111,"unitialized service: ",svname) ;
	if (!ss_state_read(&sta,ste,svname)) log_dieusys(111,"read state file of: ",ste,"/",svname) ;
	
	info_display_string(fields[5],svname) ;
	info_display_int(fields[0],sta.reload) ;
	info_display_int(fields[1],sta.init) ;
	info_display_int(fields[2],sta.unsupervise) ;
	info_display_int(fields[3],sta.state) ;
	info_display_int(fields[4],sta.pid) ;
	
	if (res.logger && logger)
	{
		svname = res.sa.s + res.logger ;
		if (!ss_state_check(ste,svname)) log_dieusys(111,"unitialized: ",svname) ;
		if (!ss_state_read(&sta,ste,svname)) log_dieusys(111,"read state file of: ",ste,"/",svname) ;
		
		if (buffer_putsflush(buffer_1,"\n") == -1)
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
		
		info_display_string(fields[6],res.sa.s + res.logreal) ;
		info_display_int(fields[0],sta.reload) ;
		info_display_int(fields[1],sta.init) ;
		info_display_int(fields[2],sta.unsupervise) ;
		info_display_int(fields[3],sta.state) ;
		info_display_int(fields[4],sta.pid) ;
	
	}

	freed:
	ss_resolve_free(&res) ;
	stralloc_free(&satree) ;
	stralloc_free(&src) ;
	stralloc_free(&tmp) ;
	return 0 ;
}
